File: 백준/Gold/2056. 작업/작업.cc
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
int w[10001];
int ind[10001];
int now[10001];
int main() {
	int n;
	cin >> n;
	vector<vector<int>> v(n + 1);
	queue<int> q;
	for (int i = 1; i <= n; i++) {
		int a, b;
		cin >> a >> b;
		w[i] = a;
		ind[i] += b;
		for (int j = 0; j < b; j++) {
			int c;
			cin >> c;
			v[c].push_back(i);
		}
	}
	int ans = 0;
	for (int i = 1; i <= n; i++) {
		if (!ind[i]) {
			q.push(i);
			now[i] += w[i];
			ans = max(ans, now[i]);
		}
	}
	while (!q.empty()) {
		int a = q.front();
		q.pop();
		for (int k : v[a]) {
			now[k] = max(now[k], now[a] + w[k]);
			if (ind[k] == 1) {
				ans = max(ans, now[k]);
				q.push(k);
			}
			ind[k]--;
		}
	}
	cout << ans;
}

File: 백준/Gold/2056. 작업/작업.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws Exception {
		var br = new BufferedReader(new InputStreamReader(System.in));

		int n = Integer.parseInt(br.readLine());
		int[] time = new int[n + 1];
		int[] in = new int[n + 1];
		List<List<Integer>> v = new ArrayList<>();
		for (int i = 0; i <= n; i++) {
			v.add(new ArrayList<>());
		}
		for (int i = 1; i <= n; i++) {
			var st = new StringTokenizer(br.readLine());
			int t = Integer.parseInt(st.nextToken());
			time[i] = t;
			int m = Integer.parseInt(st.nextToken());
			for (int j = 0; j < m; j++) {
				v.get(Integer.parseInt(st.nextToken())).add(i);
				in[i]++;
			}
		}

		int ans = 0;
		Queue<int[]> q = new ArrayDeque<>();
		for (int i = 1; i <= n; i++) {
			if (in[i] == 0) {
				q.offer(new int[]{i, time[i]});
			}
		}

		int[] max = new int[n + 1];
		while (!q.isEmpty()) {
			var now = q.poll();

//			System.out.println(now[0] + " " + now[1]);
			ans = Math.max(ans, now[1]);
			max[now[0]] = Math.max(max[now[0]], now[1]);

			for (var next : v.get(now[0])) {
				max[next] = Math.max(max[next], now[1]);
				if (--in[next] == 0) {
					q.offer(new int[]{next, max[next] + time[next]});
				}
			}
		}

		System.out.println(ans);
	}
}